src/coreComponents/dataRepository/WrapperBase.hpp

```diff
@@ -548,11 +550,67 @@ class WrapperBase
   virtual std::type_info const & getTypeId() const = 0;
 
   /**
-   * @brief Return the number of dimensions of the array if T is an array, and 0 otherwise
+   * @brief Downcast @p this to a typed wrapper.
+   * @tparam T wrapped type to cast to
+   * @pre @p this must be in instance of Wrapper<T>
+   * @return reference to @p this cast to a typed wrapper.
+   */
+  template< typename T >
+  Wrapper< T > & cast()
+  {
+    GEOSX_ERROR_IF( getTypeId() != typeid( T ), "WrapperBase: invalid downcast" );
+    return dynamicCast< Wrapper< T > & >( *this );
```

Comment:
What I want to do (and am using extensively in this PR) is
```c++
WrapperBase & wrapper = ... // get the wrapper from the group
dispatch( wrapper->getTypeId(), [&]( auto value )
{
  using T = decltype( value );
  auto & wrapperT = wrapper.cast< T >; // previously, this would've been a longer expression
  // use the wrapperT to access the stored value
} );
```

I don't think there is a way to do this without forward declaring `Wrapper< T >` in `WrapperBase.hpp`. Sure, it seems like a circular dependency, but it isn't, and there is no include cycle (the user ultimately just includes `Wrapper.hpp` once which pulls in `WrapperBase.hpp`, and that's it).

Sure, I could force the user to type
```c++
auto & wrapperT = wrapper.cast< Wrapper< T > >; 
```
but that seems like boilerplate, because `Wrapper` is the only thing that inherits from `WrapperBase`, so no other type could possibly be the target of the cast.

Reply:
> Sure, it seems like a circular dependency, but it isn't

We are here in the case where `Wrapper` depends on `WrapperBase` with inheritance and `WrapperBase` depends on `Wrapper` though `WrapperBase::cast`.
You manage to get it work because of the flexibility of C++ on this specific point.
This is a cycle and I think that in a language like `java` this would not even compile.
Moreover, having a base class know about its derived class really has the smell of wrong design here.

> there is no include cycle (the user ultimately just includes Wrapper.hpp once which pulls in WrapperBase.hpp, and that's it).

There is no include cycle because you forward declare. Otherwise, 
- To know about `WrapperBase`, `Wrapper.hpp` needs to include `WrapperBase.hpp`.
- To know about `Wrapper`, `WrapperBase.hpp` needs to include `Wrapper.hpp`.

That the developer just includes `Wrapper.hpp` is OK, but the semantic cycle is behind the scenes.

Now, could it be possible to have a nested static function (or simple free function `wrapperCast`) inside of `Wrapper` that would cast for you?
Calling `Wrapper< T >::cast( wrapperBase )` instead of `wrapperBase.cast< T >()` would do the same (at a boilerplate cost that seems reasonable).
What do you think of this option?

Reply:
This used to be in `Wrapper.hpp` but someone took it out because it wasn't being used at the time.

```
  /**
   * static function to cast a Wrapper base to a derived Wrapper<T>
   * @param base
   * @return casted Wrapper<T>
   */
  static Wrapper< T > * cast( WrapperBase * const base )
  {
#ifdef USE_DYNAMIC_CASTING
    return dynamic_cast< Wrapper< T > * >(base);
#else
    return static_cast< Wrapper< T > * >(base);
#endif
  }

  /**
   * static function to cast a Wrapper base to a derived Wrapper<T>
   * @param base
   * @return casted reference to const Wrapper<T>
   */
  static Wrapper< T > const * cast( WrapperBase const * const base )
  {
#ifdef USE_DYNAMIC_CASTING
    return dynamic_cast< Wrapper< T > const * >(base);
#else
    return static_cast< Wrapper< T > const * >(base);
#endif
  }

  /**
   * static function to cast a Wrapper base to a derived Wrapper<T>
   * @param base
   * @return casted Wrapper<T>
   */
  static Wrapper< T > & cast( WrapperBase & base )
  {
    if( base.get_typeid() != typeid(T) )
    {
      GEOS_ERROR( "invalid cast attempt" );
    }
    return static_cast< Wrapper< T > & >(base);
  }

  /**
   * static function to cast a Wrapper base to a derived Wrapper<T>
   * @param base
   * @return casted reference to const Wrapper<T>
   */
  static Wrapper< T > const & cast( WrapperBase const & base )
  {
    if( base.get_typeid() != typeid(T) )
    {
      GEOS_ERROR( "invalid cast attempt" );
    }
    return static_cast< Wrapper< T > const & >(base);
  }
```

Reply:
Pretty sure that someone was me.

Reply:
Sounds good, I will place the (reference) downcast in `Wrapper< T>`, if there are no objections